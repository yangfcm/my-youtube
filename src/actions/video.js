import { defAxios as axios } from "../settings";
import { FETCH_VIDEOS, FETCH_VIDEO, CATCH_ERROR } from "./types";
import { DEFAULT_ERROR_MSG } from "./default-error-msg";

/**
 * Get a set of videos based on filter,
 * e.g. videos belonging to a channel or videos belonging to a play list etc.
 */
export const fetchVideos = (filter, pageToken) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          ...filter,
          part: "snippet,statistics",
          maxResults: 15,
          pageToken,
        },
      });
      dispatch({
        type: FETCH_VIDEOS,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: {
          ...e.response.data.error,
          displayMessage: DEFAULT_ERROR_MSG.FAILED_TO_FETCH_VIDEO,
        },
      });
    }
  };
};

/**
 * Get a video by id
 */
export const fetchVideo = (videoId) => {
  return async (dispatch) => {
    try {
      const response = await axios.get("/videos", {
        params: {
          ...axios.defaults.params,
          part: "snippet,statistics",
          id: videoId,
        },
      });
      dispatch({
        type: FETCH_VIDEO,
        payload: response.data,
      });
    } catch (e) {
      dispatch({
        type: CATCH_ERROR,
        payload: {
          ...e.response.data.error,
          displayMessage: DEFAULT_ERROR_MSG.FAILED_TO_FETCH_VIDEO,
        },
      });
    }
  };
};
