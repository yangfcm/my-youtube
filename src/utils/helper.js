/**
 * Separate every 3 number digits by comma.
 * e.g. 12345678 => 12,345,678
 * @param {number} number
 */
export const separateNumber = number => {
  let strNum = String(number);
  let arrNum = strNum.split("");
  let arr = [];
  let part = "";
  for (let i = 0; i < arrNum.length; i++) {
    let num = arrNum[arrNum.length - i - 1];
    part = num + part;
    if ((i + 1) % 3 === 0) {
      arr.unshift(part);
      part = "";
    }
  }
  if (part.length > 0) {
    arr.unshift(part);
  }
  return arr.join(",");
};
