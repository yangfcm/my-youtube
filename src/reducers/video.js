import { FETCH_VIDEOS, FETCH_VIDEO } from "../actions/types";

export const videoReducer = (state = {}, action) => {
  switch (action.type) {
    case FETCH_VIDEOS:
      return { ...state, videos: action.payload };
    case FETCH_VIDEO:
      return { ...state, video: action.payload };
    default:
      return state;
  }
};
