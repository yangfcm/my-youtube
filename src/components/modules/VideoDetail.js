import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { separateNumber } from "../../utils/helper";

const VideoDetail = ({ video }) => {
  // console.log(video);
  return (
    <div>
      <div className="mb-3">
        <h3 className="text-primary">{video.snippet.title}</h3>
        <Link to={`/channel/${video.snippet.channelId}`}>
          {video.snippet.channelTitle}
        </Link>
      </div>

      <div className="d-flex flex-column flex-sm-row justify-content-between text-muted">
        <div id="video-mega-info">
          <FontAwesomeIcon icon="user-clock" />{" "}
          {moment(video.snippet.publishedAt).format("D MMM YYYY k:mm")} |{" "}
          <FontAwesomeIcon icon="eye" />{" "}
          {separateNumber(video.statistics.viewCount)}
        </div>
        {video.statistics.likeCount && video.statistics.dislikeCount && (
          <div id="video-stat-info">
            <span className="mr-3">
              <FontAwesomeIcon icon="thumbs-up" />{" "}
              {separateNumber(video.statistics.likeCount)}{" "}
            </span>
            <span>
              <FontAwesomeIcon icon="thumbs-down" />{" "}
              {separateNumber(video.statistics.dislikeCount)}
            </span>
          </div>
        )}
      </div>
      <hr />
      <p
        id="video-description"
        dangerouslySetInnerHTML={{ __html: video.snippet.description }}
        style={{ overflow: "hidden" }}
      ></p>
    </div>
  );
};

export default VideoDetail;
