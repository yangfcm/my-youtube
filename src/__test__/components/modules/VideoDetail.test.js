import React from "react";
import { shallow } from "enzyme";
import { Link } from "react-router-dom";
import moment from "moment";
import VideoDetail from "components/modules/VideoDetail";
import { videoResponse } from "__test__/fixtures/video";
import { separateNumber } from "utils/helper";

describe("Test videoDetail component", () => {
  let wrapper;
  const video = videoResponse.items[0];
  beforeEach(() => {
    wrapper = shallow(<VideoDetail video={video} />);
  });
  afterEach(() => {
    wrapper.unmount();
  });

  it("should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render video title", () => {
    expect(wrapper.find("h3").text()).toContain(video.snippet.title);
  });

  it("should render channel title and link", () => {
    const link = wrapper.find(Link);
    expect(link.exists()).toBe(true);
    expect(link.prop("to")).toBe(`/channel/${video.snippet.channelId}`);
    expect(link.text()).toContain(video.snippet.channelTitle);
  });

  it("should render correct publish datetime and view count", () => {
    expect(wrapper.find("div#video-mega-info").text()).toContain(
      moment(video.snippet.publishedAt).format("D MMM YYYY k:mm")
    );
    expect(wrapper.find("div#video-mega-info").text()).toContain(
      separateNumber(video.statistics.viewCount)
    );
  });

  it("should render correct thumbs up/down count", () => {
    expect(wrapper.find("div#video-stat-info").text()).toContain(
      separateNumber(video.statistics.likeCount)
    );
    expect(wrapper.find("div#video-stat-info").text()).toContain(
      separateNumber(video.statistics.dislikeCount)
    );
  });

  it("should render video description", () => {
    expect(
      wrapper.find("#video-description").prop("dangerouslySetInnerHTML")
    ).toEqual({
      __html: video.snippet.description,
    });
  });
});
