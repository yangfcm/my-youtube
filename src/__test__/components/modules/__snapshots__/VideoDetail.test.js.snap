// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test videoDetail component should render component 1`] = `
<div>
  <div
    className="mb-3"
  >
    <h3
      className="text-primary"
    >
      Democratic National Convention: Day 1
    </h3>
    <Link
      to="/channel/UC8ZLJkQZMENy8CoS8zYczBA"
    >
      2020 Democratic National Convention
    </Link>
  </div>
  <div
    className="d-flex flex-column flex-sm-row justify-content-between text-muted"
  >
    <div
      id="video-mega-info"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="user-clock"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
       
      18 Aug 2020 12:56
       |
       
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon="eye"
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
       
      662,594
    </div>
    <div
      id="video-stat-info"
    >
      <span
        className="mr-3"
      >
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon="thumbs-up"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
         
        14,471
         
      </span>
      <span>
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon="thumbs-down"
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          swapOpacity={false}
          symbol={false}
          title=""
          transform={null}
        />
         
        8,452
      </span>
    </div>
  </div>
  <hr />
  <p
    dangerouslySetInnerHTML={
      Object {
        "__html": "Tune in on Monday, August 17 from 9-11 PM ET for the first night of the 2020 Democratic National Convention, where weâ€™ll hear from the many Americans who are rising up to take on the crises facing our country.

Featuring: Former First Lady Michelle Obama, Sen. Bernie Sanders, Sen. Catherine Cortez Masto, Gov. Andrew Cuomo, Gov. Gretchen Whitmer, Rep. Jim Clyburn, Convention Chairman Rep. Bennie Thompson, Rep. Gwen Moore, Former Gov. John Kasich, Sen. Doug Jones, and Sen. Amy Klobuchar.

Follow the Democratic National Convention online:
Facebook: https://www.facebook.com/DemConvention
Twitter: https://twitter.com/DemConvention
Instagram: https://www.instagram.com/demconvention
Website: https://www.demconvention.com",
      }
    }
    id="video-description"
    style={
      Object {
        "overflow": "hidden",
      }
    }
  />
</div>
`;
